lib/constructs/identity-provider.ts

```diff
@@ -0,0 +1,247 @@
+}
+
+/**
+ * A Cognito User Pool Client-based IdP Client.
+ *
+ * This wraps the User Pool Client such that the necessary attributes are made
+ * available in a generic way.
+ */
+export class CognitoIdentityProviderClient extends Construct {
+  public readonly clientId: string;
+  public readonly name: string;
+  public readonly secret: secrets.ISecret;
+
+  constructor(scope: Construct, id: string, props: CognitoIdentityProviderClientProps) {
+    super(scope, id);
+    const client = props.userPool.addClient("Client", {
+      oAuth: {
+        flows: {
+          clientCredentials: true,
+          authorizationCodeGrant: false,
+          implicitCodeGrant: false,
+        },
+        scopes: props.scopes.map(cognito.OAuthScope.custom),
+      },
+      generateSecret: true,
+    });
+    const clientDetails = new UserPoolClientDetails(this, "ClientDetails", {
+      client,
+      userPool: props.userPool,
+    });
+    this.name = clientDetails.getResponseField("UserPoolClient.ClientName");
+    this.clientId = client.userPoolClientId;
+    this.secret = new secrets.Secret(this, "ClientSecret", {
+      secretStringValue: cdk.SecretValue.resourceAttribute(
+        clientDetails.getResponseField("UserPoolClient.ClientSecret")
+      ),
+    });
+  }
+}
+
+export interface ScopeOption {
+  name: string;
+  description: string;
+}
+
+export interface ScopeConfiguration {
+  resourceServerName?: string;
+  scopes: ScopeOption[];
+}
+
+export interface IdentityProviderProps {
+  scopes?: ScopeConfiguration[];
+}
+
+/**
+ * A generic Identity Provider application.
+ */
+export interface IIdentityProvider {
+  readonly domainName: string;
+  addClient(app: IIdentityProviderClientApplication, scopes: string[]): IIdentityProviderClient;
+}
+
+/**
+ * Required properties for creating an Identity Provider.
+ */
+export interface CognitoIdentityProviderProps extends IdentityProviderProps {
+  domainPrefix?: string;
+}
+
+/**
+ * An internal identity provider for the application.
+ */
+export class CognitoIdentityProvider extends Construct implements IIdentityProvider {
+  public readonly domainName: string;
+
+  public readonly userPool: cognito.IUserPool;
+  public readonly domain: cognito.IUserPoolDomain;
+
+  private readonly resourceServers: cognito.UserPoolResourceServer[] = [];
+  constructor(scope: Construct, id: string, props?: CognitoIdentityProviderProps) {
+    super(scope, id);
+    this.userPool = new cognito.UserPool(this, "IdP", {});
+    this.domain = this.userPool.addDomain("Domain", {
+      cognitoDomain: {
+        domainPrefix: props?.domainPrefix ?? id.toLowerCase(),
+      },
+    });
+    this.domainName = this.domain.domainName;
+    if (props?.scopes) {
+      for (const scopeConfig of props.scopes) {
+        if (!scopeConfig.resourceServerName) throw new Error("Cognito requires a resource server");
+        this.resourceServers.push(
+          this.userPool.addResourceServer(scopeConfig.resourceServerName, {
```

Comment:
I'm making sure I understand how Resource Servers work, but does this give every user in the User Pool all the scopes?

Reply:
Digging into the docs a bit more, it looks like this is what we need: https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_cognito.OAuthSettings.html

Reply:
I'm not entirely sure that I understand? A Resource Server has a name and scopes configured. Then clients can have scopes they're allowed to use; however, those scopes must have first been configured on/with a resource server associated with the pool.

I do think it's important to emphasize that in our flow, we shouldn't have users at any point, right? It's all going to be `client_credentials` flows? So the scope(s) of a user should be irrelevant.

The client can request (and be configured for) zero or more scopes on zero or more resource servers.

That object (`OAuthSettings`) is what we're passing to the second parameter of `addResourceServer`. We get a little lazy in how we define things for clients (and use use a string formatted as `<resourceServerId>/<scopeName>`) rather than requiring them to pass the whole object in.

If we don't configure a Client to have a scope here, then Cognito will not allow us to request it when we get a token. The only way to configure a scope for a client is if a Resource Server with that scope is associated with the User Pool. The Resource Server name is explicitly required to be part of the scope name as well.

So we do a little bit of `map` magic to provide a slightly nicer and more generic interface to consumers of this module, but it's all designed with these constraints in mind.

Reply:
Yes, you're right about users. I'm ultimately trying to make sure that we can configure a client with the appropriate scope, but it looks like we just need to add a `scope` param to our `/token` request to ensure that the token being used by a particular Lambda gets the least privileged scope (per standard OAuth).  Sorry for sowing some confusion here.

Reply:
Great, thanks!

Reply:
Uhh oops:

> If the client doesn't request any scopes, the authentication server uses all custom scopes associated with the client.

https://docs.aws.amazon.com/cognito/latest/developerguide/token-endpoint.html#post-token-request-parameters-in-body

Reply:
My general idea is that no client will ever be configured to be allowed more than one scope anyway? So in practice, it _shouldn't matter_. For completeness, yes, a client should still request the specific scope they need for a particular action.

So the User Pool will have all the scopes defined. Each client will have the maximal scope(s) it needs, keeping that maximum set as minimal as possible. Each client will also, at runtime, only requests the scopes it needs.